prelude/toolchains/cxx.bzl

```diff
@@ -22,52 +22,68 @@ load("@prelude//cxx:headers.bzl", "HeaderMode")
+        asm_compiler_type = default_toolchain.asm_compiler_type if ctx.attrs.compiler_type == None else ctx.attrs.compiler_type,
+        rc_compiler = default_toolchain.rc_compiler if ctx.attrs.rc_compiler == None else ctx.attrs.rc_compiler,
+        cvtres_compiler = default_toolchain.cvtres_compiler if ctx.attrs.cvtres_compiler == None else ctx.attrs.cvtres_compiler,
+        archiver = default_toolchain.archiver,
+        archiver_type = default_toolchain.archiver_type,
+        linker = default_toolchain.linker if ctx.attrs.linker == None else ctx.attrs.linker,
+        linker_type = default_toolchain.linker_type,
+        os = default_toolchain.os,
+    )
 
 def _system_cxx_toolchain_impl(ctx: AnalysisContext):
     """
     A very simple toolchain that is hardcoded to the current environment.
     """
-    archiver_args = ["ar"]
-    archiver_type = "gnu"
-    archiver_supports_argfiles = True
-    asm_compiler = ctx.attrs.compiler
-    asm_compiler_type = ctx.attrs.compiler_type
-    compiler = ctx.attrs.compiler
-    cxx_compiler = ctx.attrs.cxx_compiler
-    cvtres_compiler = ctx.attrs.cvtres_compiler
-    rc_compiler = ctx.attrs.rc_compiler
-    linker = ctx.attrs.linker
-    linker_type = "gnu"
-    pic_behavior = PicBehavior("supported")
-    binary_extension = ""
-    object_file_extension = "o"
-    static_library_extension = "a"
-    shared_library_name_default_prefix = "lib"
-    shared_library_name_format = "{}.so"
-    shared_library_versioned_name_format = "{}.so.{}"
-    additional_linker_flags = []
-    if host_info().os.is_macos:
-        archiver_supports_argfiles = False
-        linker_type = "darwin"
-        pic_behavior = PicBehavior("always_enabled")
-    elif host_info().os.is_windows:
-        msvc_tools = ctx.attrs.msvc_tools[VisualStudio]
-        archiver_args = [msvc_tools.lib_exe]
-        archiver_type = "windows"
-        asm_compiler = msvc_tools.ml64_exe
-        asm_compiler_type = "windows_ml64"
-        if compiler == "cl.exe":
-            compiler = msvc_tools.cl_exe
-        cxx_compiler = compiler
-        if cvtres_compiler == "cvtres.exe":
-            cvtres_compiler = msvc_tools.cvtres_exe
-        if rc_compiler == "rc.exe":
-            rc_compiler = msvc_tools.rc_exe
-        if linker == "link.exe":
-            linker = msvc_tools.link_exe
-        linker = _windows_linker_wrapper(ctx, linker)
+
+    toolchain_info = ctx.attrs.toolchain_info[SystemCxxToolchainInfo]
+    toolchain_info = _legacy_equivalent_toolchain_info_windows(ctx, toolchain_info) if toolchain_info.os == "windows" else _legacy_equivalent_toolchain_info_non_windows(ctx, toolchain_info)
+
+    archiver_supports_argfiles = toolchain_info.os != "macos"
+    additional_linker_flags = ["-fuse-ld=lld"] if toolchain_info.os == "linux" and toolchain_info.linker != "g++" and toolchain_info.cxx_compiler != "g++" else []
+
+    if toolchain_info.os == "windows":
```

Comment:
I was already more-or-less getting this "os" attribute from `_target_os_type`, indirectly though the `SystemCxxToolchainInfo` (now `CxxToolsInfo`). But I did the change you requested, removing the os from `CxxToolsInfo` and adding it directly to the cxx toolchain rules.
I think the change makes sense because the "os" only determines the flags and command line arguments that the `CxxToolsInfo` will run with, it is not a characteristic of the compiler/linker/CxxTool itself. Not sure if this the same reasoning as you but if the end result is the same, it doesn't matter that much.